### ISSUE TITLE
Seed pre-actions or seed order dependency

### ISSUE THREAD

**Issue by TJKoury** [tgriesser/knex#629]
One of my seeds requires another to complete first, I prefixed them correctly to run in alphanumeric order (1_name, 2_name, etc).  Unfortunately because the interface to bluebird promises within exports.seed is opaque I have no real flow control.  I'd like to mod the library to have an 'exports.pre' and 'exports.post', and/or an entry in the knexfile that specifies seed requirements.

Thoughts?

**Comment by bendrucker** [comment]
Would appreciate a real example of what you need. I've always been against the inclusion of seed feature because I don't see any hard problems to solve. I view it as a case where the need is so simple that you're better off writing your own code and solving your problem exactly right (as I do) versus trying to work around a library. Migrations are the opposite case — rewriting that from scratch would be a huge pain.

**Comment by tgriesser** [comment]
I think the problem being solved here is just standardizing a convention for a common pattern that a lot of people need to handle when working on an application. It's easy to do it yourself in your own way when a team is small, but once you start having lots of people working on one or more applications it's easier to just define a documented way of how to do something.

@TJKoury I agree, I was using seeds for the first time the other day and realized that they should come timestamped and also given the option to run them incrementally... I also agree with having more hooks into these things. We're going to start splitting out the library into different independent modules (seeds, migrations, builder, cli, etc etc) and they will be versioned and maintained individually, but still all  `require`'ed and exposed from a single `knex` module.

Once we start down that path, it'll be easier to identify what and where things need to be added to each piece.

**Comment by TJKoury** [comment]
So specifically I was trying to add a 'cars' table with a field for each car containing a random uuid id of an actual seeded user.  The issue is wrapping the promise returned from a db call, and/or lacking a clean way to pass data between seed instances to be called within scope of each Promise.

What I have below works, but I had to basically wrap a Promise in a Promise then call Promise.settle() on a list of dynamically generated Promises.  Not ideal, but did the job.

I'm certainly open to other ways to accomplish this.

```
'use strict';
var uuid = require('uuid');
var _Promise = require('bluebird');
var entity = "cars";

exports.seed = function(knex, Promise) {

  return Promise.all([new _Promise(function() {

    knex('users').then(function (data) {

      for (var i = 0; i < 20000; i++) {
        var user = data[Math.floor(Math.random() * (data.length - 1) + 1)];

        Promise.settle([knex(entity).insert({
          'id': uuid.v4(),
          'name': null,
          'user_id': user.id,
          'created_at': new Date(),
          'updated_at': new Date()
        })]);
      }

    })
  })]);

}

```

**Comment by tgriesser** [comment]
Hmm, I don't see why you'd need a separate copy of bluebird? Also, it seems you're using promises incorrectly... you should be able to accomplish the same with this:

```js
'use strict';
var uuid     = require('uuid');
var entity   = "cars";

exports.seed = function(knex, Promise) {
  return knex('users').then(function(data) {
    var inserts = []
    for (var i = 0; i < 20000; i++) {
      inserts.push({
        'id': uuid.v4(),
        'name': null,
        'user_id': data[Math.floor(Math.random() * (data.length - 1) + 1)].id,
        'created_at': new Date(),
        'updated_at': new Date()
      });
    }
    return knex.insert(inserts).into(entity);
  })
}
```

**Comment by tgriesser** [comment]
Basically, you shouldn't ever have promises that aren't `return`'ed, otherwise you're likely using them wrong.

**Comment by TJKoury** [comment]
@tgriesser while I agree with you theoretically, that code you posted (as well as plenty of my own permutations) never returns.  If I return a 'knex promise' it just hangs forever, and nothing is ever written to the database.  I'll keep screwing with it and post if I find something that works and is cleaner.

**Comment by TJKoury** [comment]
Ok, here's a cleaner version that works, still not ideal:

```

var uuid = require('uuid');
var entity = "cars";


exports.seed = function(knex, Promise) {
  return Promise.all([knex('users').then(function (data) {
      var inputs = [];
      for (var i = 0; i < 20000; i++) {
        var user = data[Math.floor(Math.random() * (data.length - 1) + 1)];

        inputs.push(knex(entity).insert({
          'id': uuid.v4(),
          'name': null,
          'user_id': user.id,
          'created_at': new Date(),
          'updated_at': new Date()
        }));
      }
      return Promise.settle(inputs);
  })]);

}

```

**Comment by bendrucker** [comment]
Using `Promise.settle` here is wrong. It's designed to wait for all promises to either be resolved or rejected and return inspection objects. You don't want that. If the insert fails it should fail immediately. You're also generating 20,000 separate queries whereas Tim's version is sending only one.

**Comment by bendrucker** [comment]
I'm still not understanding what we could add here to make this better for you. If you wrote your own script you could easily store state and pass it around as a context with `bind`.

**Comment by TJKoury** [comment]
@tgriesser @bendrucker Thanks for getting back to me so quick with this, I do appreciate the response time.

@bendrucker I agree completely and want to run it as one query, but this hangs:

```
'use strict';

var uuid = require('uuid');
var entity = "cars";

exports.seed = function(knex, Promise) {
  return knex('users').then(function (data) {
      var inputs = [];
      for (var i = 0; i < 20000; i++) {
        var user = data[Math.floor(Math.random() * (data.length - 1) + 1)];

        inputs.push({
          'id': uuid.v4(),
          'name': null,
          'user_id': user.id,
          'created_at': new Date(),
          'updated_at': new Date()
        });
      }
      return knex.insert(inputs).into(entity);
  });
}

```

Even if I try to return it wrapped in Promise.all(knex.insert(inputs).into(entity)) , it hangs forever.  That knex promise that gets returned doesn't seem to work in this context.

As far as what you can do for me, this is fine, I might submit a pull sometime in the future once this particular task is done that handles this functionality a bit cleaner if you don't mind.  

Writing my own scripts against frameworks is something I stay away from when working on a team, unless there is a documented extensibility mechanism.  If I could write my own knex commands to extend (like artisan) that would be neat.  If you can already do that please forgive my cursory reading of the docs.

**Comment by tgriesser** [comment]
How long are you waiting for it to complete... I mean you are inserting 20,000 records in a single statement, that could take a bit - the parsing for insert statements needs to be improved internally. Does the same code hang if you change 20,000 to 100?

Even so there's no need to call `Promise.all` in either case

**Comment by TJKoury** [comment]
@tgriesser it's negligible, a few seconds at most with the code I posted that's working ^^^.  With the hanging version it never returns even with one, and the database isn't even being hit, something is fubar internally. 

Node 10.35, OSX: 10.9.5 (13F34), PostgreSQL 9.4.0.1 (with postgis extension, but no extension column types on this table).

**Comment by bendrucker** [comment]
https://github.com/bendrucker/knex-issue-629

The short of it is that there's no answer other than to break up the inserts into chunks right now. I'm surprised no one's run into this before. Inserting roughly 1,000 records works fine. It takes almost a second. 

Once the workload reaches the 5,000 range on my machine, it leaps to 16 seconds. This is what happens when Node runs at 100% CPU. GC and the event loop fight with each other and everything's a mess.

**Comment by TJKoury** [comment]
@bendrucker Thanks for that due diligence, after further testing I can confirm that's the behavior I'm seeing as well.  It might be a good idea to introduce a 'batch' setting that allows manual adjustment of batch size.

**Comment by bendrucker** [comment]
Really just needs further investigation into what calls are eating all the CPU time. It's an odd use case since if you should be streaming data of that size in rather than trying to assemble a massive query. Can't see how 20,000 records would be necessary for testing.

**Comment by TJKoury** [comment]
I guess everyone has different use cases, I deal with millions of records
on a regular basis (and usually hand-tune the SQL), so 20,000 really isn't
much.  For now batching seems to work just fine so I'm a happy camper.

You mentioned streaming but I didn't see support for the postgis COPY
command, is there something I'm missing?

Really just needs further investigation into what calls are eating all the
CPU time. It's an odd use case since if you should be streaming data of
that size in rather than trying to assemble a massive query. Can't see how
20,000 records would be necessary for testing.

--
Reply to this email directly or view it on GitHub
<https://github.com/tgriesser/knex/issues/629#issuecomment-70380574>.

**Comment by bendrucker** [comment]
https://github.com/brianc/node-pg-copy-streams

**Comment by TJKoury** [comment]
I'm aware of it, is there a hook into knex?

**Comment by bendrucker** [comment]
No

**Comment by TJKoury** [comment]
Thanks again for all the help.  You guys have done great work, and in my case has helped bring over a few folks to Node from the Laravel world.

**Comment by bendrucker** [comment]
Very glad to hear it! If you have specific suggestions on how to make the seed API work better for you in feel free to open a new issue so it's easier to track down.
